stargz/stargz.go

```diff
@@ -401,9 +459,26 @@ func (fr *fileReader) ReadAt(p []byte, off int64) (n int, err error) {
 	//  offset by the chunk's offset.
 	off -= ent.ChunkOffset
 
+	finalEnt := fr.ents[len(fr.ents)-1]
 	gzOff := ent.Offset
-	sr := io.NewSectionReader(fr.r.sr, gzOff, fr.r.sr.Size()-gzOff)
-	gz, err := gzip.NewReader(sr)
+	// gzBytesRemain is the number of compressed gzip bytes in this
+	// file remaining, over 1+ gzip chunks.
+	gzBytesRemain := finalEnt.NextOffset() - gzOff
+
+	sr := io.NewSectionReader(fr.r.sr, gzOff, gzBytesRemain)
+
+	const maxGZread = 2 << 20
+	var bufSize int = maxGZread
+	if gzBytesRemain < maxGZread {
+		bufSize = int(gzBytesRemain)
+	}
+
+	br := bufio.NewReaderSize(sr, bufSize)
+	if _, err := br.Peek(bufSize); err != nil {
```

Comment:
What's the purpose of this `Peek`? Just probing to fail fast?

Reply:
By default the bufio.NewReader will return a reader that reads only 4KB at a time. And because its underlying reader is ultimately the read-range-requests-from-gcr.io Reader, that'll result in lots of 4KB HTTP requests. This pre-slurps into memory the rest of this gzip chunk (say, 1MB or 4MB) so the gzip reader doesn't end up doing a bazillion HTTP requests.